gusto/initialisation_tools.py

```diff
@@ -293,117 +307,181 @@ def moist_hydrostatic_balance(state, theta_e, water_t, pi_boundary=Constant(1.0)
     # Calculate hydrostatic Pi
     Vt = theta0.function_space()
     Vr = rho0.function_space()
-    Vv = state.fields('u').function_space()
-    n = FacetNormal(state.mesh)
-    g = state.parameters.g
-    cp = state.parameters.cp
-    R_d = state.parameters.R_d
-    p_0 = state.parameters.p_0
 
     VDG = state.spaces("DG")
     if any(deg > 2 for deg in VDG.ufl_element().degree()):
         state.logger.warning("default quadrature degree most likely not sufficient for this degree element")
-    quadrature_degree = (5, 5)
-
-    params = {'ksp_type': 'preonly',
-              'ksp_monitor_true_residual': True,
-              'ksp_converged_reason': True,
-              'snes_converged_reason': True,
-              'ksp_max_it': 100,
-              'mat_type': 'aij',
-              'pc_type': 'lu',
-              'pc_factor_mat_solver_type': 'mumps'}
 
     theta0.interpolate(theta_e)
     water_v0.interpolate(water_t)
-    Pi = Function(Vr)
-    epsilon = 0.9  # relaxation constant
-
-    # set up mixed space
-    Z = MixedFunctionSpace((Vt, Vt))
-    z = Function(Z)
-
-    gamma, phi = TestFunctions(Z)
-
-    theta_v, w_v = z.split()
-
-    # give first guesses for trial functions
-    theta_v.assign(theta0)
-    w_v.assign(water_v0)
-
-    theta_v, w_v = split(z)
-
-    # define variables
-    T = thermodynamics.T(state.parameters, theta_v, Pi, r_v=w_v)
-    p = thermodynamics.p(state.parameters, Pi)
-    w_sat = thermodynamics.r_sat(state.parameters, T, p)
-
-    dxp = dx(degree=(quadrature_degree))
-
-    # set up weak form of theta_e and w_sat equations
-    F = (-gamma * theta_e * dxp
-         + gamma * thermodynamics.theta_e(state.parameters, T, p, w_v, water_t) * dxp
-         - phi * w_v * dxp
-         + phi * w_sat * dxp)
-
-    problem = NonlinearVariationalProblem(F, z)
-    solver = NonlinearVariationalSolver(problem, solver_parameters=params)
-
-    theta_v, w_v = z.split()
 
-    Pi_h = Function(Vr).interpolate((p / p_0) ** (R_d / cp))
+    rho_h = Function(Vr)
+    rho_averaged = Function(Vt)
+    rho_broken = Function(FunctionSpace(state.mesh, BrokenElement(Vt.ufl_element())))
+    rho_recoverer = Recoverer(rho_broken, rho_averaged)
+    w_h = Function(Vt)
+    theta_h = Function(Vt)
+    theta_e_test = Function(Vt)
+    delta = 0.8
+
+    # expressions for finding theta0 and water_v0 from theta_e and water_t
+    pie = thermodynamics.pi(state.parameters, rho_averaged, theta0)
+    p = thermodynamics.p(state.parameters, pie)
+    T = thermodynamics.T(state.parameters, theta0, pie, water_v0)
+    r_v_expr = thermodynamics.r_sat(state.parameters, T, p)
+    theta_e_expr = thermodynamics.theta_e(state.parameters, T, p, water_v0, water_t)
+
+    iterations = 0
```

Comment:
Do you need the `iterations` variable? Can't you just check the value of `i`after the for loop?